libcudacxx/include/cuda/std/inplace_vector

```diff
@@ -0,0 +1,1746 @@
+  {
+    return static_cast<size_type>(this->__size_);
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr bool empty() const noexcept
+  {
+    return this->__size_ == 0;
+  }
+
+  // [containers.sequences.inplace.vector.data], data access
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr _Tp* data() noexcept
+  {
+    return _CUDA_VSTD::launder(reinterpret_cast<_Tp*>(this->__elems_));
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const _Tp* data() const noexcept
+  {
+    return _CUDA_VSTD::launder(reinterpret_cast<const _Tp*>(this->__elems_));
+  }
+
+  // [containers.sequences.inplace.vector.itertators] iterators
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY iterator begin() noexcept
+  {
+    return _CUDA_VSTD::launder(reinterpret_cast<_Tp*>(this->__elems_));
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY const_iterator begin() const noexcept
+  {
+    return _CUDA_VSTD::launder(reinterpret_cast<const _Tp*>(this->__elems_));
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY iterator end() noexcept
+  {
+    return _CUDA_VSTD::launder(reinterpret_cast<_Tp*>(this->__elems_) + this->__size_);
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY const_iterator end() const noexcept
+  {
+    return _CUDA_VSTD::launder(reinterpret_cast<const _Tp*>(this->__elems_) + this->__size_);
+  }
+
+  // [containers.sequences.inplace.vector.modifiers], modifiers
+  _LIBCUDACXX_INLINE_VISIBILITY void __destroy(iterator __first, iterator __last) noexcept
+  {
+    _CUDA_VSTD::__destroy(__first, __last);
+    __change_size(__first - __last);
+  }
+
+  template <class... _Args>
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr reference
+  unchecked_emplace_back(_Args&&... __args) noexcept(_CCCL_TRAIT(is_nothrow_constructible, _Tp, _Args...))
+  {
+    auto __final = _CUDA_VSTD::__construct_at(end(), _CUDA_VSTD::forward<_Args>(__args)...);
+    this->__change_size(1);
+    return *__final;
+  }
+
+  _LIBCUDACXX_TEMPLATE(bool _IsNothrow = _CCCL_TRAIT(is_nothrow_default_constructible, _Tp))
+  _LIBCUDACXX_REQUIRES(_IsNothrow)
+  _LIBCUDACXX_INLINE_VISIBILITY void __uninitialized_value_construct(iterator __first, iterator __last) noexcept
+  {
+    iterator __idx = __first;
+    for (; __idx != __last; ++__idx)
+    {
+      ::new (_CUDA_VSTD::__voidify(*__idx)) _Tp();
+    }
+    __change_size(__last - __first);
+  }
+
+  _LIBCUDACXX_TEMPLATE(bool _IsNothrow = _CCCL_TRAIT(is_nothrow_default_constructible, _Tp))
+  _LIBCUDACXX_REQUIRES((!_IsNothrow))
+  _LIBCUDACXX_INLINE_VISIBILITY void __uninitialized_value_construct(iterator __first, iterator __last)
+  {
+    iterator __idx = __first;
+    auto __guard   = __make_exception_guard(_Rollback_change_size<__inplace_vector_base>{this, __first, __idx});
+    for (; __idx != __last; ++__idx)
+    {
+      ::new (_CUDA_VSTD::__voidify(*__idx)) _Tp();
+    }
+    __guard.__complete();
+    __change_size(__last - __first);
+  }
```

Comment:
Question: why do we need a guard to rollback the size, if we only change the size after the potentially throwing region is finished (after the guard complets)?

Wouldn't we rather need to destroy the already constructed elements here?

Reply:
That is exactly the point with the basic exception guarantee. 

We are required to keep all elements intact that have been constructed. So when an exception is throw the rollback will update the size of the container to reflect the new elements.

`std::uninitialized_meow` would clean up all previous elements when an exception is thrown

Reply:
I think I understand. Thx!